File: algos/sorting/counting_sort.py
def counting_sort(L):
    n = len(L)

    pos = [0] * n
    for num in L:
        pos[num] += 1

    num_count = n
    for i in reversed(range(n)):
        pos[i] = num_count - pos[i]
        num_count = pos[i]

    sorted_L = [None] * n
    for i in range(n):
        item = L[i]
        new_idx = pos[item]
        sorted_L[new_idx] = item
        pos[item] += 1

    return sorted_L

if __name__ == "__main__":
    l = [2, 3, 4, 3, 1]
    new_l = counting_sort(l)
    print(new_l)

File: algos/sorting/countingSort.ts
const countingSort = (arr: number[]): number[] => {
  const n = arr.length;

  // initialize pos to [0, 0, 0] (length n)
  const pos: number[] = new Array(n).fill(0)

  // arr = [1, 2, 1]
  // pos = [0, 2, 1]
  for (const num of arr) {
    pos[num] += 1
  }

  // count = 3
  // pos = [0, 2, 1]
  // pos' = [0, 0, 2]
  let count: number = n
  for (let i = n - 1; i >= 0; i--) {
    pos[i] = count - pos[i]
    count = pos[i]
  }

  // [1, 2, 1] => [1, 1, 2]
  const sortedArr: number[] = []
  for (const num of arr) {
    const new_idx: number = pos[num]
    sortedArr[new_idx] = num
    pos[num] += 1
  }

  return sortedArr
}

const startArr = [4, 2, 1, 5, 3, 3]
console.log(countingSort(startArr))
